### PR TITLE
Modifications for Creator UI parser

### PR DESCRIPTION
**Pull request by kalinon**
Ive pulled the parser files and put them into a separate shard [qt5-ui](https://github.com/kalinon/qt5-ui.cr), it will be easier to maintain that way. This PR just contains the adjustments to `bindgen` config.

---

Ive added an initial parser to use QT Creator UI files to build the GUI. This parses the XML file and performs the heavy lifting in getting the GUI up and running. This is just a starting point and i will add more supported types as i run into them. Need #46 to be merged before.

Example files were added in: samples/creator_ui/

- xml: samples/creator_ui/mainwindow.ui
- crystal: samples/creator_ui/main.cr

Creator GUI layout:
image

Generated crystal UI:
image

### CODE REVIEW

**Comment by Papierkorb** [comment]
Oh, that looks like a cool addition!  I know that plenty people prefer using an UI editor over writing it in code (Although I find the Crystal API much more snappy than the C++ API heh).

What are your plans with this PR?

**Comment by kalinon** [comment]
> Oh, that looks like a cool addition! I know that plenty people prefer using an UI editor over writing it in code (Although I find the Crystal API much more snappy than the C++ API heh).
> 
> What are your plans with this PR?

We can either bring it into the qt5 shard, or i can create a sub shard. Figured id throw it out there and see what you thought.

File: qt.yml [hunk 1]
```diff
@@ -10,7 +10,10 @@
 # These bindings will, as configured, bind to Qt dynamically.
 
 module: Qt
-library: "%/ext/binding_{BINDING_PLATFORM}.a -lstdc++ -lQt5Core -lQt5Gui -lQt5Widgets"
+if_os_is_darwin:
+  library: "%/ext/binding_{BINDING_PLATFORM}.a -lstdc++ `pkg-config --libs Qt5Widgets`"
```
Comment (Papierkorb) [inline qt.yml]: We can use `pkg-config` on linux as well

File: support/generate_bindings.cr [hunk 2]
```diff
@@ -11,19 +11,24 @@ require "ini"
...
-#      OS       LIBC   ARCH      Qt     Clang target triplet      Ptr  Endian
-#  { "linux", "gnu", "x86_64", "5.5",  "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.6",  "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.7",  "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.8",  "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.9",  "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.10", "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.11", "x86_64-unknown-linux-gnu", 8, "little" },
-#  { "linux", "gnu", "x86_64", "5.12", "x86_64-unknown-linux-gnu", 8, "little" },
-  { "linux", "gnu", "x86_64", "5.13", "x86_64-unknown-linux-gnu", 8, "little" },
```
Comment (Papierkorb) [inline support/generate_bindings.cr:24]: No linuxes at all?  Also, I think we can get rid of anything prior Qt5.9

File: src/qt5/ui/parser.cr [hunk 3]
```diff
@@ -0,0 +1,85 @@
+require "xml"
+require "spoved/logger"
```
Comment (Papierkorb) [inline src/qt5/ui/parser.cr]: What library is that?

Comment (kalinon) [reply src/qt5/ui/parser.cr]: this is the logging lib i use to shortcut making `logger` methods and bindings. This is imported via `bindgen`

**Comment by kalinon** [comment]
Converted this to a draft as its not quite ready yet. but its going well.

File: qt.yml (hunk 1 above)
Comment (kalinon) [reply qt.yml]: i can do that

File: support/generate_bindings.cr (hunk 2 above)
Comment (kalinon) [reply support/generate_bindings.cr:24]: Ill drop things before Qt5.9

**Review by kalinon — COMMENTED** [review]
made requested changes

**Comment by kalinon** [comment]
@Papierkorb Ive pulled the parser files and put them into a separate shard [qt5-ui](https://github.com/kalinon/qt5-ui.cr), it will be easier to maintain that way. This PR just contains the adjustments to `bindgen` config.

File: ext/bindgen_helper.hpp [hunk 4]
```diff
@@ -39,11 +39,12 @@
 
 #ifndef BINDGEN_HELPER_HPP
 #define BINDGEN_HELPER_HPP
+#define GC_NEW_ABORTS_ON_OOM
```
Comment (Papierkorb) [inline ext/bindgen_helper.hpp:42]: I don't even remember anymore why this file is mirrored from Bindgen.  In any case, could you mirror the changes over?  That define sounds like a good idea!

File: src/qt5/abstract_list_model.cr [hunk 5]
```diff
@@ -0,0 +1,13 @@
+abstract class Qt::AbstractListModel
+  def row_count : Int32
```
Comment (Papierkorb) [inline src/qt5/abstract_list_model.cr:2]: Aren't `#row_count` and `#column_count` auto-generated by Bindgen?

File: README.md [hunk 6]
```diff
@@ -68,7 +68,7 @@ on this under License below.)
 Using your system's package manager:
 
 * **ArchLinux** `pacman -S qt5-base`
-* **Debian** `apt-get install qtbase5-dev`
+* **Debian** `brew install qt5`
```
Comment (Papierkorb) [inline README.md]: Brew on Debian?  You can also add MacOS in that list :)

If you want, you can work on/add a chapter on using your shard.  I think that would make it much easier to find to the curious tinkerer.  Would be great if you could update the README on your end as well (Some TODOs etc)

**Review by Papierkorb — CHANGES_REQUESTED** [review]
Good call with the split into another shard.  Small-ish changes, nothing major.

File: README.md (hunk 6 above)
Comment (kalinon) [reply README.md]: No this was be reading Debian as **Darwin**... ill fix it up.

Comment (kalinon) [reply README.md]: but also interesting enough, there is brew for linux: https://docs.brew.sh/Homebrew-on-Linux

File: src/qt5/abstract_list_model.cr (hunk 5 above)
Comment (kalinon) [reply src/qt5/abstract_list_model.cr:2]: I thought so too, but these jump table entries dont seem to make it:

```
bg_QAbstractItemModel_rowCount_const_QModelIndex_R: BindgenHelper.wrap_proc({% if forwarded.includes?("row_count") %} Proc(Binding::QModelIndex*, Int32).new{|parent| self.row_count(ModelIndex.new(unwrap: parent)) } {% else %} nil {% end %}),
bg_QAbstractItemModel_columnCount_const_QModelIndex_R: BindgenHelper.wrap_proc({% if forwarded.includes?("column_count") %} Proc(Binding::QModelIndex*, Int32).new{|parent| self.column_count(ModelIndex.new(unwrap: parent)) } {% else %} nil {% end %}),
```

Compiler keept yelling at me till i made these entries.

Comment (Papierkorb) [reply src/qt5/abstract_list_model.cr:2]: Hmpf.  We can investigate this later I guess.

**Pull request closed without merging**